compiler/src/test/scala/org/scalajs/core/compiler/test/OptimizationTest.scala

```diff
@@ -49,6 +49,46 @@ class OptimizationTest extends JSASTTest {
     }
   }
 
+  @Test
+  def testArrayApplyOptimization: Unit = {
+    /* Make sure Array(...) is optimized away completely for several kinds
+     * of data types, with both the generic overload and the ones specialized
+     * for primitives.
+     */
+    """
+    class A {
+      val a = Array(5, 7, 9, -3)
+      val b = Array("hello", "world")
+      val c = Array('a', 'b')
+      val d = Array(Nil)
+      val e = Array(5.toByte, 7.toByte, 9.toByte, -3.toByte)
+    }
+    """.
+    hasNot("any of the Array.apply methods") {
+      case js.Apply(js.LoadModule(jstpe.ClassType("s_Array$")), js.Ident(methodName, _), _)
+          if methodName.startsWith("apply__") =>
+    }
+
+    /* Using [] with primitives produces suboptimal trees, which cannot be
+     * optimized. We should improve this in the future, if possible. This is
+     * particularly annoying for Byte and Short, as it means that we need to
+     * write `.toByte` for every single element if we want the optimization to
+     * kick in.
+     *
+     * Scala/JVM has the same limitation.
+     */
+    """
+    class A {
+      val a = Array[Int](5, 7, 9, -3)
+      val b = Array[Byte](5, 7, 9, -3)
+    }
+    """.
+    hasExactly(2, "calls to Array.apply methods") {
+      case js.Apply(js.LoadModule(jstpe.ClassType("s_Array$")), js.Ident(methodName, _), _)
+          if methodName.startsWith("apply__") =>
+    }
```

Comment:
Why do we need to test this?

Reply:
We don't need to, but I feel it's a nice way to document that limitation for ourselves, when inevitably we'll come back to this commit because we're wondering why a call to `Array()` is not optimized the way we'd like.